src/test/java/com/snowflake/kafka/connector/internal/StageFilesProcessorTest.java

```diff
@@ -0,0 +1,249 @@
+import com.snowflake.kafka.connector.internal.telemetry.SnowflakeTelemetryService;
+import net.snowflake.client.jdbc.internal.joda.time.DateTime;
+import net.snowflake.client.jdbc.internal.joda.time.DateTimeZone;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.*;
+
+
+class StageFilesProcessorTest {
+
+    private static final String PIPE_NAME = "testPipe";
+    private static final String TABLE_NAME = "testTable";
+    private static final String STAGE_NAME = "testStage";
+    private static final String PREFIX = "prefix";
+
+    private SnowflakeConnectionService conn;
+    private SnowflakeIngestionService ingestionService;
+    private AtomicLong currentTime;
+    private StageFilesProcessor victim;
+    private StageFilesProcessor.ProgressRegisterImpl register;
+    private SnowflakeTelemetryPipeCreation pipeCreation;
+    private SnowflakeTelemetryPipeStatus pipeTelemetry;
+    private SnowflakeTelemetryService telemetryService;
+
+    @BeforeEach
+    void setUp() {
+        currentTime = new AtomicLong(new DateTime(2000, 1, 10, 12, 0, DateTimeZone.UTC).getMillis());
+
+        conn = Mockito.mock(SnowflakeConnectionService.class);
+        ingestionService = Mockito.mock(SnowflakeIngestionService.class);
+        pipeCreation = new SnowflakeTelemetryPipeCreation(TABLE_NAME, STAGE_NAME, PIPE_NAME);
+        pipeTelemetry = Mockito.mock(SnowflakeTelemetryPipeStatus.class);
+        telemetryService = Mockito.mock( SnowflakeTelemetryService.class);
+        register = new StageFilesProcessor.ProgressRegisterImpl(pipeCreation, pipeTelemetry, telemetryService);
+
+        victim = new StageFilesProcessor(PIPE_NAME, TABLE_NAME, STAGE_NAME, PREFIX, conn, ingestionService, pipeTelemetry, telemetryService, currentTime::get);
+    }
+
+    @Test
+    void filesProcessWillTerminateOnStopSignal() throws InterruptedException {
+        AtomicBoolean completed = new AtomicBoolean(false);
+        Thread worker = new Thread(() -> {
+            victim.trackFiles(register, Thread.currentThread(), () -> !Thread.currentThread().isInterrupted(), d -> Thread.sleep(10L));
+            completed.set(true);
+        });
+        worker.start();
+        Thread stopper = new Thread(() -> {
+            try {
+                Thread.sleep(Duration.ofSeconds(2L).toMillis());
+                victim.close();
+            } catch (Exception e) {
+            }
+        });
+        stopper.start();
+        stopper.join(Duration.ofSeconds(5L).toMillis());
+
+        assertThat(completed.get()).isTrue();
```

Comment:
We can use awaitility instead.
```java
await().atMost(2, SECONDS).until(completed::get);
```

Reply:
i didn't notice we have that library at hand !